File: Algorithm.Python/BasicTemplateCryptoAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from clr import AddReference
AddReference("System")
AddReference("QuantConnect.Common")

from System import *
from QuantConnect import *
from QuantConnect.Brokerages import *
from QuantConnect.Orders import *
from QCAlgorithm import QCAlgorithm

import decimal as d

### <summary>
### The demonstration algorithm shows some of the most common order methods when working with Crypto assets.
### </summary>
### <meta name="tag" content="using data" />
### <meta name="tag" content="using quantconnect" />
### <meta name="tag" content="trading and orders" />
class BasicTemplateCryptoAlgorithm(QCAlgorithm):

    def Initialize(self):
        '''Initialise the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must initialized.'''

        self.SetStartDate(2018, 4, 4)  #Set Start Date
        self.SetEndDate(2018, 4, 4)    #Set End Date

        # Although typically real brokerages as GDAX only support a single account currency,
        # here we add both USD and EUR to demonstrate how to handle non-USD account currencies.
        # Set Strategy Cash (USD)
        self.SetCash(10000)

        # Set Strategy Cash (EUR)
        # EUR/USD conversion rate will be updated dynamically
        self.SetCash("EUR", 10000, 1.23)

        # Add some coins as initial holdings
        # When connected to a real brokerage, the amount specified in SetCash
        # will be replaced with the amount in your actual account.
        self.SetCash("BTC", 1, 7300)
        self.SetCash("ETH", 5, 400)

        # Note: the conversion rates above are required in backtesting (for now) because of this issue:
        # https://github.com/QuantConnect/Lean/issues/1859

        self.SetBrokerageModel(BrokerageName.GDAX, AccountType.Cash)

        # You can uncomment the following lines when live trading with GDAX,
        # to ensure limit orders will only be posted to the order book and never executed as a taker (incurring fees).
        # Please note this statement has no effect in backtesting or paper trading.
        # self.DefaultOrderProperties = GDAXOrderProperties()
        # self.DefaultOrderProperties.PostOnly = True

        # Find more symbols here: http://quantconnect.com/data
        self.AddCrypto("BTCUSD", Resolution.Minute)
        self.AddCrypto("ETHUSD", Resolution.Minute)
        self.AddCrypto("BTCEUR", Resolution.Minute)
        symbol = self.AddCrypto("LTCUSD", Resolution.Minute).Symbol

        # create two moving averages
        self.fast = self.EMA(symbol, 30, Resolution.Minute)
        self.slow = self.EMA(symbol, 60, Resolution.Minute)

    def OnData(self, data):
        '''OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.

        Arguments:
            data: Slice object keyed by symbol containing the stock data
        '''

        # Note: all limit orders in this algorithm will be paying taker fees,
        # they shouldn't, but they do (for now) because of this issue:
        # https://github.com/QuantConnect/Lean/issues/1852

        if self.Time.hour == 1 and self.Time.minute == 0:
            # Sell all ETH holdings with a limit order at 1% above the current price
            limitPrice = round(self.Securities["ETHUSD"].Price * d.Decimal(1.01), 2)
            quantity = self.Portfolio.CashBook["ETH"].Amount
            self.LimitOrder("ETHUSD", -quantity, limitPrice)

        elif self.Time.hour == 2 and self.Time.minute == 0:
            # Submit a buy limit order for BTC at 5% below the current price
            usdTotal = self.Portfolio.CashBook["USD"].Amount
            limitPrice = round(self.Securities["BTCUSD"].Price * d.Decimal(0.95), 2)
            # use only half of our total USD
            quantity = usdTotal * d.Decimal(0.5) / limitPrice
            self.LimitOrder("BTCUSD", quantity, limitPrice)

        elif self.Time.hour == 2 and self.Time.minute == 1:
            # Get current USD available, subtracting amount reserved for buy open orders
            usdTotal = self.Portfolio.CashBook["USD"].Amount
            usdReserved = sum(x.Quantity * x.LimitPrice for x
                in [x for x in self.Transactions.GetOpenOrders()
                    if x.Direction == OrderDirection.Buy
                        and x.Type == OrderType.Limit
                        and (x.Symbol.Value == "BTCUSD" or x.Symbol.Value == "ETHUSD")])
            usdAvailable = usdTotal - usdReserved
            self.Debug("usdAvailable: {}".format(usdAvailable))

            # Submit a marketable buy limit order for ETH at 1% above the current price
            limitPrice = round(self.Securities["ETHUSD"].Price * d.Decimal(1.01), 2)

            # use all of our available USD
            quantity = usdAvailable / limitPrice

            # this order will be rejected (for now) because of this issue:
            # https://github.com/QuantConnect/Lean/issues/1852
            self.LimitOrder("ETHUSD", quantity, limitPrice)

            # use only half of our available USD
            quantity = usdAvailable * d.Decimal(0.5) / limitPrice
            self.LimitOrder("ETHUSD", quantity, limitPrice)

        elif self.Time.hour == 11 and self.Time.minute == 0:
            # Liquidate our BTC holdings (including the initial holding)
            self.SetHoldings("BTCUSD", 0)

        elif self.Time.hour == 12 and self.Time.minute == 0:
            # Submit a market buy order for 1 BTC using EUR
            self.Buy("BTCEUR", 1)

            # Submit a sell limit order at 10% above market price
            limitPrice = round(self.Securities["BTCEUR"].Price * d.Decimal(1.1), 2)
            self.LimitOrder("BTCEUR", -1, limitPrice)

        elif self.Time.hour == 13 and self.Time.minute == 0:
            # Cancel the limit order if not filled
            self.Transactions.CancelOpenOrders("BTCEUR")

        elif self.Time.hour > 13:
            # To include any initial holdings, we read the LTC amount from the cashbook
            # instead of using Portfolio["LTCUSD"].Quantity

            if self.fast > self.slow:
                if self.Portfolio.CashBook["LTC"].Amount == 0:
                    self.Buy("LTCUSD", 10)
            else:
                if self.Portfolio.CashBook["LTC"].Amount > 0:
                    # The following two statements currently behave differently if we have initial holdings:
                    # https://github.com/QuantConnect/Lean/issues/1860

                    self.Liquidate("LTCUSD")
                    # self.SetHoldings("LTCUSD", 0)

    def OnOrderEvent(self, orderEvent):
        self.Debug("{} {}".format(self.Time, orderEvent.ToString()))

    def OnEndOfAlgorithm(self):
        self.Log("{} - TotalPortfolioValue: {}".format(self.Time, self.Portfolio.TotalPortfolioValue))
        self.Log("{} - CashBook: {}".format(self.Time, self.Portfolio.CashBook))

File: Algorithm.FSharp/BasicTemplateCryptoAlgorithm.fs
(*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*)


namespace QuantConnect.Algorithm.FSharp

open System
open QuantConnect
open QuantConnect.Algorithm
open QuantConnect.Orders
open QuantConnect.Brokerages
open Interop
open QuantConnect.Interfaces

// Declare algorithm name
type BasicTemplateCryptoAlgorithm() =

    //Reuse all the base class of QCAlgorithm
    inherit QCAlgorithm()

    let mutable _fast = Unchecked.defaultof<_>
    let mutable _slow = Unchecked.defaultof<_>

    override this.Initialize() =

        this.SetStartDate(2018, 4, 4) // Set Start Date
        this.SetEndDate(2018, 4, 4) // Set End Date

        // Although typically real brokerages as GDAX only support a single account currency,
        // here we add both USD and EUR to demonstrate how to handle non-USD account currencies.
        // Set Strategy Cash (USD)
        this.SetCash(10000)

        // Set Strategy Cash (EUR)
        // EUR/USD conversion rate will be updated dynamically
        this.SetCash("EUR", 10000m, 1.23m)

        // Add some coins as initial holdings
        // When connected to a real brokerage, the amount specified in SetCash
        // will be replaced with the amount in your actual account.
        this.SetCash("BTC", 1m, 7300m)
        this.SetCash("ETH", 5m, 400m)

        // Note: the conversion rates above are required in backtesting (for now) because of this issue:
        // https://github.com/QuantConnect/Lean/issues/1859

        this.SetBrokerageModel(BrokerageName.GDAX, AccountType.Cash)

        // You can uncomment the following line when live trading with GDAX,
        // to ensure limit orders will only be posted to the order book and never executed as a taker (incurring fees).
        // Please note this statement has no effect in backtesting or paper trading.
        // DefaultOrderProperties = new GDAXOrderProperties { PostOnly = true }

        // Find more symbols here: http://quantconnect.com/data
        this.AddCrypto("BTCUSD") |> ignore
        this.AddCrypto("ETHUSD") |> ignore
        this.AddCrypto("BTCEUR") |> ignore
        let symbol = this.AddCrypto("LTCUSD").Symbol

        // create two moving averages
        _fast <- this.EMA(symbol, 30, Nullable Resolution.Minute)
        _slow <- this.EMA(symbol, 60, Nullable Resolution.Minute)

    override this.OnData data =
        match this.Time.Hour, this.Time.Minute with
        | (1,0) ->
            let limitPrice = Math.Round(this.Securities.["ETHUSD"].Price * 1.01m, 2)
            let quantity = this.Portfolio.CashBook.["ETH"].Amount
            this.LimitOrder(!>"ETHUSD", -quantity, limitPrice) |> ignore

        | (2,0) -> 
            let usdTotal = this.Portfolio.CashBook.["USD"].Amount
            let limitPrice = Math.Round(this.Securities.["BTCUSD"].Price * 0.95m, 2)
            // use only half of our total USD
            let quantity = usdTotal * 0.5m / limitPrice
            this.LimitOrder(!> "BTCUSD", quantity, limitPrice) |> ignore

        | (2,1) -> 
            let usdTotal = this.Portfolio.CashBook.["USD"].Amount
            let usdReserved = 
                this.Transactions.GetOpenOrders(fun x -> x.Direction = OrderDirection.Buy && x.Type = OrderType.Limit)
                |> Seq.cast<LimitOrder>
                |> Seq.filter (fun x-> x.Symbol = !> "BTCUSD" || x.Symbol = !> "ETHUSD")
                |> Seq.sumBy (fun x -> x.Quantity * x.LimitPrice)
            let usdAvailable = usdTotal - usdReserved

            // Submit a marketable buy limit order for ETH at 1% above the current price
            let limitPrice = Math.Round(this.Securities.["ETHUSD"].Price * 1.01m, 2)

            // use all of our available USD
            let quantity = usdAvailable / limitPrice

            // this order will be rejected for insufficient funds
            this.LimitOrder(!>"ETHUSD", quantity, limitPrice) |> ignore

            // use only half of our available USD
            let quantity = usdAvailable * 0.5m / limitPrice
            this.LimitOrder(!> "ETHUSD", quantity, limitPrice) |> ignore

        | (11,0) ->
            this.SetHoldings(!>"BTCUSD", 0m)

        | (12,0) ->
            this.Buy( !> "BTCEUR", 1m) |> ignore
            // Submit a sell limit order at 10% above market price
            let limitPrice = Math.Round(this.Securities.["BTCEUR"].Price * 1.1m, 2)
            this.LimitOrder(!>"BTCEUR", -1, limitPrice) |> ignore

        | (13,0) -> 
            this.Transactions.CancelOpenOrders(!>"BTCEUR") |> ignore

        | (t,_) when t > 13 ->
            match _fast > _slow, this.Portfolio.CashBook.["LTC"].Amount with
            | true, 0m              -> this.Buy((!> "LTCUSD":Symbol), 10) |> ignore
            | false, a when a > 0m  -> this.Liquidate(!> "LTCUSD") |>  ignore
            | _                     -> ()

        | (_,_)  -> ()                  
                  
    override this.OnOrderEvent orderEvent = 
        this.Debug(sprintf "%A %A" this.Time orderEvent)

    override this.OnEndOfAlgorithm() =
        this.Log(sprintf "%A - TotalPortfolioValue:%A" this.Time this.Portfolio.TotalPortfolioValue)


    interface IRegressionAlgorithmDefinition with
        member __.CanRunLocally = true

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        member __.Languages = [| Language.CSharp; Language.Python; Language.FSharp |]

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        member __.ExpectedStatistics = 
            idict [
                "Total Trades", "10"
                "Average Win", "0%"
                "Average Loss", "-0.18%"
                "Compounding Annual Return", "-99.992%"
                "Drawdown", "3.800%"
                "Expectancy", "-1"
                "Net Profit", "-2.545%"
                "Sharpe Ratio", "-16.028"
                "Loss Rate", "100%"
                "Win Rate", "0%"
                "Profit-Loss Ratio", "0"
                "Alpha", "-5.47"
                "Beta", "326.539"
                "Annual Standard Deviation", "0.201"
                "Annual Variance", "0.04"
                "Information Ratio", "-16.112"
                "Tracking Error", "0.2"
                "Treynor Ratio", "-0.01"
                "Total Fees", "$85.27"
            ]

    